### ISSUE TITLE
Support `MaybeUninit`

### ISSUE THREAD

**Issue by QuarticCat** [rsmpi/rsmpi#152]
Currently, `all_reduce_into` and some other APIs only accept buffers like `&mut [f32]` but not `&mut [MaybeUninit<f32>]`. However, It's common to use `&mut [MaybeUninit<f32>]` as a receive buffer. We need lib support for this, as casting a uninitialized buffer to `&mut [T]` might be undefined behavior, see [[1]](https://rust-lang.github.io/rfcs/2930-read-buf.html), [[2]](https://github.com/rust-lang/unsafe-code-guidelines/issues/71), and [[3]](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.slice_assume_init_mut).

Note that we cannot simply implement `Equivalence` for `MaybeUninit<T>` where `T: Equivalence<Out = SystemDatatype>`, which gives users the right to pass an uninitialized buffer as a send buffer, so that APIs like `all_reduce_into` are not safe anymore.

**Comment by jedbrown** [comment]
I agree this would be useful in some form, though there are issues. There is related discussion in #133 that we should consider.

**Comment by jedbrown** [comment]
So the straightforward user interface might look like
```rs
// this is safe because `assume_init` applies to the `MaybeUninit<f64>`, but you get the buffer somehow
let mut buffer: [MaybeUninit<f64>; 4] = unsafe { MaybeUninit::uninit().assume_init() };

let recv: &mut [MaybeUninit<f64>] = &mut buffer[..]; // slice to receive into
comm.all_reduce_into(send, recv, op);
let recv: &mut [f64] = unsafe { MaybeUninit::slice_assume_init_mut(recv) }; // nightly API or can be done with transmute
```

I think that's pretty undesirable because that unsafe line at the bottom is easy for callers to mess up (and a bit worse without using the [nightly-only experimental API](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.slice_assume_init_mut)).

What about this?
```rs
let recv: &mut [MaybeUninit<f64>] = &mut buffer[..]; // slice to receive into
let recv: &mut [f64] = comm.all_reduce_into_uninit(send, recv, op);
```

Note that I used overly type annotations above, and in practice, you'd just write:
```rs
let recv = comm.all_reduce_into_uninit(send, &mut buffer[..], op);
```

**Comment by QuarticCat** [comment]
I've left the original project that needed this feature for a long time and forgot some details. But the `all_reduce_into_uninit` API looks good to me.

**Comment by jedbrown** [comment]
This post has more discussion of safe buffer interfaces (for rustix). It's worth reading before implementing our own interface. 
https://blog.sunfishcode.online/writingintouninitializedbuffersinrust/

**Comment by jedbrown** [comment]
`BorrowedBuf` as in [`read_buf`](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_buf) is an alternative interface, though we would need to make our own generic version (preferably in a separate crate).

A [proposal](https://github.com/rust-lang/libs-team/issues/258) to make `BorrowedBuf` generic was rejected.
